db/model/subject.js

```diff
@@ -539,37 +542,122 @@ module.exports = function subject(seq, dataTypes) {
+                // if it does not exist, throw 404 subject not found
+                // if subject exists, does parent id === parentId?
+                // if true, return parent
+                return new seq.Promise((resolve, reject) => {
+                  if (inst.parentAbsolutePath === inst.absolutePath) {
+                    throw new IllegalSelfParenting({
+                      message: 'absolutePath cannot equal parentAbsolutePath: ' + value,
+                    });
+                  }
+
+                  Subject.findOne({
+                    where: { absolutePath: value },
+                  })
+                  .then((parent) => {
+                    // if parentId changed too, check for consistency
+                    if (inst.changed('parentId') && inst.parentId) {
+
+                      if (parent && parent.id !== value1) {
+                        // parentAbsolutePath does not match parentId
+                        throw new ParentSubjectNotMatch({
+                          message: value + ' does not match ' + value1,
+                        });
+                      } else if (!parent) {
+                        // no parent found
+                        throw new ParentSubjectNotFound({
+                          message: value + ' not found.',
+                        });
+                      }
+                    }
+
+                    // if parents match, or if only parentAbsolutePath changed
+                    // return parent
+                    resolve(parent);
+                  })
+                  .catch((err) => {
+                    reject(err);
+                  });
+                });
+              } else if (!inst.changed('parentId')) {
+
+                // body contains parentAbsolutePath with empty value
+                // set parentAbsolutePath to null
+                return new seq.Promise((resolve, reject) => {
+                  resolve();
+                });
+              }
+            } {
+              // only the parentId field changed. It may be empty
+              // find parent subject
+              if (inst.parentId === inst.id) {
+                throw new IllegalSelfParenting({
+                  message: 'id cannot equal parentId: ' + value,
+                });
+              }
+
+              setValues();
+              return Subject.scope({ method: [param, value] }).find();
+            }
           }
 
           return new seq.Promise((resolve, reject) => {
-            Subject.scope({ method: [param, value] }).find()
+            checkParent()
             .then((parent) => {
               if (parent) {
                 inst.setDataValue('absolutePath',
                   parent.absolutePath + '.' + inst.name);
                 inst.setDataValue(key, value);
-                inst.setDataValue(key1, parent.getDataValue(param1));
+                inst.setDataValue(key1, parent[param1]);
```

Comment:
This code must be working, but it was very confusing for me to understand. I feel this code can be improved a lot in terms of readability when using specific variables instead of setValues function.